HIVE-25576: Configurable datetime formatter for unix_timestamp, from_unixtime

zabetak opened the pull request:
### What changes were proposed in this pull request?
1. Add hive.datetime.formatter property to control formatter in unix_timestamp and from_unixtime functions.
2. Add UnixTimeFormatter class hierarchy for encapsulating formatting and parsing of unixtime based on the configuration.
3. Refactor unix_timestamp (+vectorized) and from_unixtime implementations to use the new formatter classes.
4. Add parameterized unit tests for the affected UDF implementations.

### Why are the changes needed?
The two Java formatters present diffences in their behavior leading to different query results. The supported patterns, between the two formatters, are also different something that makes existing queries crash at runtime (after upgrade). Adapting to the new behavior of DateTimeFormatter is a challenging and time-consuming task for end users especially due to the widespread use of the afforementioned unixtime functions.

Although DateTimeFormatter is a clear improvement over SimpleDateFormat some users still want to retain the old behavior for compatibility reasons thus introducing a property is necessary for facilitating migration.

### Does this PR introduce _any_ user-facing change?
Yes, adds a new property that users can tune to change the behavior of `unix_timestamp` and `from_unixtime` functions.

### Is the change a dependency upgrade?
No

### How was this patch tested?
New + existing tests
```
mvn test -pl ql -Dtest=TestGenericUDFFromUnixTimeEvaluate,TestVectorUDFUnixTimeStampString,TestGenericUDFToUnixTimestampEvaluateStringString
mvn test -pl itests/qtest -Dtest=TestMiniLlapLocalCliDriver -Dqfile_regex=.*unix.* -Pitests
```

amansinha100 commented on ql/src/java/org/apache/hadoop/hive/ql/udf/generic/UnixTimeFormatter.java:
```diff
@@ -0,0 +1,77 @@
+  enum Type {
+    SIMPLE {
+      @Override
+      UnixTimeFormatter<?> newFormatter(ZoneId zone) {
```
nit:  add final for the zone to be consistent with the DATETIME newFormatter method.

zabetak replied:
Instead of adding, I opted to remove the final (just adds noise in this case) from the other method. https://github.com/apache/hive/pull/4615/commits/d8e02d58895613197aa3190e3c64c890a601228f

amansinha100 commented on ql/src/java/org/apache/hadoop/hive/ql/udf/generic/UnixTimeFormatter.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  protected final ZoneId zoneId;
+  private final LoadingCache<String, T> formatters;
```
In the typical case, do we expect the cache to be populated with more than 2 instances ?  I was thinking if it would be sufficient to keep 2 separate member variables for each of the formatters. If we absolutely need the LoadingCache , a comment to justify would be useful.

amansinha100 replied:
Yeah, that's why I felt the LoadingCache was more than what's needed in this case. Simplification is good.

jfsii commented:
Besides @amansinha100 comments, LGTM. Nice work on the the tests. AFAICT the test failure is likely a flaky test.

zabetak commented:
@jfsii One test failure was indeed a bug I introduced and fixed in https://github.com/apache/hive/pull/4615/commits/bc19ca28fd088ae1662b1771525a2a02a698c302.

jfsii commented:
> @jfsii One test failure was indeed a bug I introduced and fixed in [bc19ca2](https://github.com/apache/hive/commit/bc19ca28fd088ae1662b1771525a2a02a698c302).

Doh! Looking back at the prior run I was looking at the wrong view and only noticed the compactor failure and not the obviously related failure :/

amansinha100 reviewed (commented):
Changes LGTM.  +1

saihemanth-cloudera reviewed (approved):
LGTM +1, Pending green Jenkins tests.
